packages/itwin/tree-widget/src/components/trees/stateless/common/components/VisibilityTree.tsx

```diff
@@ -3,26 +3,44 @@
  * See LICENSE.md in the project root for license terms and full copyright notice.
  *--------------------------------------------------------------------------------------------*/
 
-import { useMemo } from "react";
+import { useCallback, useEffect, useMemo, useState } from "react";
 import { useReportingAction } from "../../../common/UseFeatureReporting";
 import { useHierarchyVisibility } from "../UseHierarchyVisibility";
 import { useMultiCheckboxHandler } from "../UseMultiCheckboxHandler";
+import { createIModelAccess } from "../Utils";
 import { BaseTree } from "./BaseTree";
 import { TreeRenderer } from "./TreeRenderer";
 
+import type { ECClassHierarchyInspector } from "@itwin/presentation-shared";
+import type { HierarchyVisibilityHandler } from "../UseHierarchyVisibility";
 import type { ComponentPropsWithoutRef } from "react";
 
 type BaseTreeProps = ComponentPropsWithoutRef<typeof BaseTree>;
 type UseHierarchyVisibilityProps = Parameters<typeof useHierarchyVisibility>[0];
-type VisibilityTreeProps = Omit<BaseTreeProps, "treeRenderer"> & UseHierarchyVisibilityProps;
+type VisibilityTreeProps = Omit<BaseTreeProps, "treeRenderer" | "imodelAccess"> &
+  Omit<UseHierarchyVisibilityProps, "visibilityHandlerFactory"> & {
+    visibilityHandlerFactory: (imodelAccess: ECClassHierarchyInspector) => HierarchyVisibilityHandler;
+  };
 
 /** @internal */
 export function VisibilityTree({ visibilityHandlerFactory, ...props }: VisibilityTreeProps) {
-  const { getCheckboxState, onCheckboxClicked } = useHierarchyVisibility({ visibilityHandlerFactory });
+  const { imodel, getSchemaContext } = props;
+  const imodelAccess = useMemo(() => createIModelAccess({ imodel, getSchemaContext }), [imodel, getSchemaContext]);
+  const { getCheckboxState, onCheckboxClicked, triggerRefresh } = useHierarchyVisibility({
+    visibilityHandlerFactory: useCallback(() => visibilityHandlerFactory(imodelAccess), [visibilityHandlerFactory, imodelAccess]),
+  });
+  const [onRootNodesLoaded, setOnRootNodesLoaded] = useState<() => void>();
+
+  useEffect(() => {
+    // When filtered paths change, refresh checkboxes just after the filtered hierarchy is loaded.
+    setOnRootNodesLoaded(() => () => triggerRefresh());
+  }, [triggerRefresh, props.getFilteredPaths]);
```

Comment:
I think we can always refresh visibility state when root nodes are loaded (tree reloaded). You can simplify and pass `onRootNodesLoaded={triggerRefresh}` to `BaseTree`.

Reply:
Made it so that it always calls `triggersRefresh` on tree reload